File: scripts/migrate-sermons.js
#!/usr/bin/env node

const fs = require("fs");
const path = require("path");

const feedDirectory = process.argv[2] || "/tmp";
const outputPath = path.resolve(process.argv[3] || "content/sermons.json");
const feedFiles = fs.readdirSync(feedDirectory)
  .filter((name) => /^grasso-feed(?:-\d+)?\.xml$/.test(name))
  .sort((a, b) => {
    const page = (name) => Number(name.match(/-(\d+)\.xml$/)?.[1] || 1);
    return page(a) - page(b);
  });

const books = [
  "Song of Solomon", "1 Thessalonians", "2 Thessalonians", "1 Corinthians", "2 Corinthians",
  "1 Chronicles", "2 Chronicles", "Deuteronomy", "Ecclesiastes", "Lamentations", "Philippians",
  "Colossians", "Revelation", "1 Samuel", "2 Samuel", "1 Timothy", "2 Timothy", "1 Peter", "2 Peter",
  "1 Kings", "2 Kings", "1 John", "2 John", "3 John", "Leviticus", "Numbers", "Joshua", "Judges",
  "Nehemiah", "Esther", "Psalms", "Proverbs", "Isaiah", "Jeremiah", "Ezekiel", "Daniel", "Hosea",
  "Obadiah", "Habakkuk", "Zephaniah", "Zechariah", "Matthew", "Romans", "Galatians", "Ephesians",
  "Philemon", "Hebrews", "Genesis", "Exodus", "Ruth", "Ezra", "Job", "Joel", "Amos", "Jonah",
  "Micah", "Nahum", "Haggai", "Malachi", "Mark", "Luke", "John", "Acts", "James", "Titus", "Jude"
];

function decodeEntities(value) {
  const named = { amp: "&", apos: "'", quot: '"', lt: "<", gt: ">", nbsp: " " };
  return value
    .replace(/&#(\d+);/g, (_, number) => String.fromCodePoint(Number(number)))
    .replace(/&#x([0-9a-f]+);/gi, (_, number) => String.fromCodePoint(parseInt(number, 16)))
    .replace(/&([a-z]+);/gi, (entity, name) => named[name.toLowerCase()] ?? entity);
}

function textContent(value) {
  return decodeEntities(value)
    .replace(/<!\[CDATA\[|\]\]>/g, "")
    .replace(/<script[\s\S]*?<\/script>/gi, " ")
    .replace(/<style[\s\S]*?<\/style>/gi, " ")
    .replace(/<[^>]+>/g, " ")
    .replace(/\s+/g, " ")
    .trim();
}

function tag(item, name) {
  return item.match(new RegExp(`<${name}(?:\\s[^>]*)?>([\\s\\S]*?)<\\/${name}>`, "i"))?.[1] || "";
}

function inferBook(title, description) {
  if (/^Psalm\b/i.test(title)) return "Psalms";
  if (/^Song of Songs\b/i.test(title)) return "Song of Solomon";
  if (/^1st Peter\b/i.test(title)) return "1 Peter";
  const titleMatch = books.find((book) => new RegExp(`^${book.replace(/ /g, "\\s+")}\\b`, "i").test(title));
  if (titleMatch) return titleMatch;
  const scriptureMatch = books.find((book) => new RegExp(`\\b${book.replace(/ /g, "\\s+")}\\s+\\d`, "i").test(description));
  return scriptureMatch || "";
}

const sermonsById = new Map();

for (const filename of feedFiles) {
  const xml = fs.readFileSync(path.join(feedDirectory, filename), "utf8");
  const items = xml.match(/<item>[\s\S]*?<\/item>/gi) || [];

  for (const item of items) {
    const title = textContent(tag(item, "title"));
    const sourceUrl = textContent(tag(item, "link"));
    const guid = textContent(tag(item, "guid"));
    const rawDescription = textContent(tag(item, "description"));
    const description = rawDescription.replace(/\s*The post .*? appeared first on New Covenant OPC\.?$/i, "").trim();
    const enclosure = item.match(/<enclosure\s+[^>]*url="([^"]+)"/i)?.[1] || "";
    const id = guid.match(/[?&]p=(\d+)/)?.[1] || sourceUrl.split("/").filter(Boolean).pop();
    const date = new Date(textContent(tag(item, "pubDate"))).toISOString().slice(0, 10);

    if (!title || !sourceUrl || !id) continue;
    sermonsById.set(String(id), {
      id: String(id),
      title,
      book: inferBook(title, description),
      date,
      description,
      audioUrl: decodeEntities(enclosure),
      sourceUrl
    });
  }
}

const sermons = [...sermonsById.values()].sort((a, b) => b.date.localeCompare(a.date) || b.id.localeCompare(a.id));
fs.mkdirSync(path.dirname(outputPath), { recursive: true });
fs.writeFileSync(outputPath, `${JSON.stringify({ sermons }, null, 2)}\n`);

const missingBooks = sermons.filter(({ book }) => !book).length;
console.log(`Migrated ${sermons.length} sermons from ${feedFiles.length} feed pages to ${outputPath}.`);
console.log(`${sermons.filter(({ audioUrl }) => audioUrl).length} include MP3 audio; ${missingBooks} need a book assigned manually.`);

File: scripts/migrate-sermons.rb
#!/usr/bin/env ruby

require "cgi"
require "date"
require "fileutils"
require "json"
require "rexml/document"

feed_directory = ARGV[0] || "/tmp"
output_path = File.expand_path(ARGV[1] || "content/sermons.json")
feed_files = Dir.glob(File.join(feed_directory, "grasso-feed*.xml")).sort_by do |filename|
  File.basename(filename)[/-(\d+)\.xml$/, 1]&.to_i || 1
end

books = [
  "Song of Solomon", "1 Thessalonians", "2 Thessalonians", "1 Corinthians", "2 Corinthians",
  "1 Chronicles", "2 Chronicles", "Deuteronomy", "Ecclesiastes", "Lamentations", "Philippians",
  "Colossians", "Revelation", "1 Samuel", "2 Samuel", "1 Timothy", "2 Timothy", "1 Peter", "2 Peter",
  "1 Kings", "2 Kings", "1 John", "2 John", "3 John", "Leviticus", "Numbers", "Joshua", "Judges",
  "Nehemiah", "Esther", "Psalms", "Proverbs", "Isaiah", "Jeremiah", "Ezekiel", "Daniel", "Hosea",
  "Obadiah", "Habakkuk", "Zephaniah", "Zechariah", "Matthew", "Romans", "Galatians", "Ephesians",
  "Philemon", "Hebrews", "Genesis", "Exodus", "Ruth", "Ezra", "Job", "Joel", "Amos", "Jonah",
  "Micah", "Nahum", "Haggai", "Malachi", "Mark", "Luke", "John", "Acts", "James", "Titus", "Jude"
]

def plain_text(value)
  CGI.unescapeHTML(value.to_s.gsub(/<[^>]+>/, " ").gsub(/\s+/, " ").strip)
end

sermons_by_id = {}

feed_files.each do |filename|
  document = REXML::Document.new(File.read(filename))
  REXML::XPath.each(document, "//item") do |item|
    title = plain_text(item.elements["title"]&.text)
    source_url = plain_text(item.elements["link"]&.text)
    guid = plain_text(item.elements["guid"]&.text)
    description = plain_text(item.elements["description"]&.text)
      .sub(/\s*The post .*? appeared first on New Covenant OPC\.?$/i, "")
      .strip
    audio_url = CGI.unescapeHTML(item.elements["enclosure"]&.attributes&.fetch("url", "").to_s)
    id = guid[/[?&]p=(\d+)/, 1] || source_url.split("/").reject(&:empty?).last
    next if title.empty? || source_url.empty? || id.to_s.empty?

    book_aliases = {
      /^Psalm\b/i => "Psalms",
      /^Song of Songs\b/i => "Song of Solomon",
      /^1st Peter\b/i => "1 Peter"
    }
    book = book_aliases.find { |pattern, _name| title.match?(pattern) }&.last
    book ||= books.find { |name| title.match?(/^#{Regexp.escape(name)}\b/i) }
    book ||= books.find { |name| description.match?(/\b#{Regexp.escape(name)}\s+\d/i) }

    sermons_by_id[id.to_s] = {
      id: id.to_s,
      title: title,
      book: book || "",
      date: DateTime.parse(plain_text(item.elements["pubDate"]&.text)).strftime("%Y-%m-%d"),
      description: description,
      audioUrl: audio_url,
      sourceUrl: source_url
    }
  end
rescue REXML::ParseException => error
  warn "Skipped #{filename}: #{error.message}"
end

sermons = sermons_by_id.values.sort_by { |sermon| [sermon[:date], sermon[:id]] }.reverse
FileUtils.mkdir_p(File.dirname(output_path))
File.write(output_path, JSON.pretty_generate({ sermons: sermons }) + "\n")

puts "Migrated #{sermons.length} sermons from #{feed_files.length} feed pages to #{output_path}."
puts "#{sermons.count { |sermon| !sermon[:audioUrl].empty? }} include MP3 audio; #{sermons.count { |sermon| sermon[:book].empty? }} need a book assigned manually."
